### PR TITLE
[WIP] Adding the ability to filter reads by Bismap mappability

### PR DESCRIPTION
**Pull request by Valiec**
I have seen that some aligners (e.g. Bismark) align reads with high MapQ where they shouldn't (e.g. centromeres), leading to incorrect methylation calling. Using data from Bismap (https://bismap.hoffmanlab.org/), it is possible to filter these out with bedtools, but such positional filtering is too strict.  Reads having a mate in a mappable region are uniquely placed and can be used even if the read is in a non-unique location.

To rescue these reads, I've created a new feature for MethylDackel to filter read pairs by Bismap mappability score. This checks if either mate of a read pair is aligned in a unique enough region (according to Bismap data) to support methylation calls, and if not, discards the entire read pair. Thus we avoid methylation calls in non-unique regions without sacrificing calls near mappable regions. 

Specifically, it looks at each read pair to see if a certain number of bases (default: 15) have a Bismap mappability score that is high enough (default: >=0.01), and if not, discards the read pair.

It is invoked with the -M or -B options, and the min number of mappable bases and mappability threshold are configurable with -b and -t respectively.

With -M, this takes a bigWig file, which it loads using libBigWig (this library is now a dependency for building the tool from source).

With -B, it does the same thing, but instead of a bigWig, it loads the data from a custom compressed binary format called BBM (short for Binary BisMap). See the BBM_Specification file in this pull request for a full specification on the file format. This was added because loading from a large bigWig can be slow, and a BBM (for the k=100 Bismap data file) is about 4x faster to load.

-O outputs a BBM with the same name as the input bigWig (but with the extension .bbm instead of .bw or .bigWig), and -N outputs a BBM with whatever name you specify as an argument to the option.

### CODE REVIEW

File: extract.c [hunk 1]
```diff
@@ -109,7 +111,7 @@ char revcomp(char b) {
         case 'T':
         case 't':
             return 'A';
-        default:
+
```
Comment (dpryan79) [inline extract.c]: Any reason you removed the `default:`?

File: extract.c [hunk 2]
```diff
@@ -235,7 +237,14 @@ int isVariant(Config *config, const bam_pileup1_t *plp, uint32_t *coverage, int
...
 }
 
+int error()
+{
+    printf("fatal: malformed BBM file\n");
+    return -9;
+}
+
 void *extractCalls(void *foo) {
+    //fprintf(stderr, "in extractCalls\n");
```
Comment (dpryan79) [inline extract.c]: Can you remove this?

File: main.c [hunk 3]
```diff
@@ -26,6 +26,7 @@ void usage_main() {
 }
 
 int main(int argc, char *argv[]) {
+    //putc((int)('.'), stderr);
```
Comment (dpryan79) [inline main.c]: Please remove

File: extract.c (hunk 1 above)
Comment (Valiec) [reply extract.c]: This appears to have happened by accident, will re-add.

File: extract.c (hunk 2 above)
Comment (Valiec) [reply extract.c]: Thanks, I thought I removed all of these, guess I missed some.

File: main.c (hunk 3 above)
Comment (Valiec) [reply main.c]: Thanks, I thought I removed all of these, guess I missed some.

**Comment by dpryan79** [comment]
Thanks @Valiec (and @bwlang)! I'll be back in the office next week and would like to play around with this a bit before tagging a new release.

**Pull request merged**